torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1733,6 +1736,57 @@ struct to_ir {
     emitLoopCommon(stmt.range(), emit_body, nullptr, {}, cond);
   }
 
+  void emitWith(const With& stmt) {
+    auto targets = stmt.targets();
+    // Keep a stack of entered objects so they can be exited
+    // in the right order.
+    std::stack<Value*> entered;
+
+    for (const auto& target : targets) {
+      Expr e = target.target();
+
+      auto* rhs = emitExpr(e);
+      auto* n = graph->insertNode(graph->create(prim::Enter, {rhs}));
+      entered.push(rhs);
+      auto rhsClass = rhs->type()->expect<ClassType>();
+
+      if (!rhsClass) {
+        throw ErrorReport(e.range())
+            << "With item expression does not return a class type";
+      }
+
+      auto* enterMethod = rhsClass->getMethod("__enter__");
+      auto* exitMethod = rhsClass->getMethod("__exit__");
+
+      if (!enterMethod || !exitMethod) {
+        throw ErrorReport(e.range())
+            << "Object returned by with item expression does not define __enter__ and __exit__ methods";
+      }
+
+      // Set the output of the enter node to be the return type of __enter__.
+      n->output(0)->setType(
```

Comment:
We should figure out what we're going to do about typing here. For `__enter__`, that's easy, we can check that it's a method with no arguments (although in practice no one will break that). For `__exit__` i'm not sure exactly what type signatures we should accept, but we should probably be checking something.